### PR TITLE
training script modification

### PR DESCRIPTION
**Pull request by samuelmyoussef**
Modified the `rl.py` script:

- Removed algorithms from stable baselines (SB2)
- Added the remaining algorithms from stable baselines (SB3): A2C, DQN, TD3, DDPG
- Added all the available environments in SofaGym
- Added command line arguments to choose the env, algo, and training parameters
- Added the choice to continue training from a previously trained model
- Added Tensorboard logging

Added a documentation part to briefly explain the script usage.

### CODE REVIEW

File: rl.py [hunk 1]
```diff
@@ -182,71 +320,133 @@ def sec_to_hours(seconds):
...
+               
+            model.save(f"{model_dir}/latest")
 
-               if r >= best:
-                   print(">>>    Save training n°", idx+1)
-                   model.save("./Results_benchmark/" +  name + "/best")
+        print(">>   End.")
+        print("[INFO]  >>    time: ", sec_to_hours(time.time()-start_time))
+    
+    elif TRAIN == 'continue':
```
Comment (EtienneMenager) [inline rl.py:370]: Nice to add the possibility to continue the training. Do you have a specific initialisation of the *idx* and *best* parameters?

Comment (samuelmyoussef) [reply rl.py:370]: You can choose the saved time step you want to continue from with the command line argument "-ms", but the default is the latest time step in the case of continuing the training, and the best in the case of testing the model.

I still have a problem reloading the idx from where it previously stopped. I could save it as a parameter in a separate file alongside maybe the model hyperparameters and load it back with the model, but for now it restarts from 0, but in the tensorboard logging it will show the total timesteps.

Comment (EtienneMenager) [reply rl.py:370]: Ok perfect!

**Review by EtienneMenager — APPROVED** [review]

File: rl.py [hunk 2]
```diff
@@ -32,144 +33,281 @@
...
     return d
 
+
+results_dir = "./Results_benchmark"
+models_dir = f"{results_dir}/models"
+logs_dir = f"{results_dir}/logs"
+videos_dir = f"{logs_dir}/videos"
+
+envs = {
+        1: 'bubblemotion-v0',
```
Comment (alxbilger) [inline rl.py:109]: Why `v0` at the end of all names?

File: README.md [hunk 3]
```diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
...
+RL algorithms can be used with SofaGym to train RL agents for the provided environments to perform specific tasks. Currently, the training in SofaGym is done using Stable Baselines.
+
+To train a model for one of the registered environments of SofaGym, the ```rl.py``` script can be used:
+
+```bash
+python rl.py -e trunk-v0 -a PPO
+```
+
+Command line arguments can be used to change the training parameters:
+- -e, --env: environment name from available examples
```
Comment (alxbilger) [inline README.md:117]: It could be great to be able to list the names of all available examples

File: README.md [hunk 4]
```diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
...
+- -a, --algo: RL algorithm name from the algorithms supported by stable baselines
```
Comment (alxbilger) [inline README.md:118]: It could be great to be able to list the names of all available algorithms

File: README.md [hunk 5]
```diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
...
+- -ne, --env_num: number of environments trained in parallel [Optional, default=4]
+- -ep, --epochs: number of training epochs [Optional, default=100]
+- -tr, --train: option to choose between training a new model, continue training of an available model, or no training [Optional, default=new]
+- -te, --te: option to test the trained model [Optional, default=False]
+- -tn, --num_test: number of tests to run on the trained model [Optional, default=1]
+- -md, --model_dir: path to save the model training files
+- -ms, --model_step: time step to load a trained model from
+
```
Comment (alxbilger) [inline README.md:126]: I see some parameters from the script are missing from the documentation

File: rl.py [hunk 6]
```diff
@@ -32,144 +33,281 @@
...
+    n_epochs = args.epochs
+    n_steps = args.steps
+    batch_size = args.batch
+    lr = args.learn_rate
+    gamma = args.gamma
+    TRAIN = args.train
+    TEST = args.test
+    n_test = args.num_test
+
+    layers_size = [512, 512, 512]
```
Comment (alxbilger) [inline rl.py:190]: It is usually not a good practice to use hard coded values. But it is maybe out of the scope of this PR

File: rl.py [hunk 7]
```diff
@@ -32,144 +33,281 @@
...
+    os.makedirs(log_dir, exist_ok=True)
+    os.makedirs(video_dir, exist_ok=True)
 
 
-    gamma = gammas[args.num_env]
-    learning_rate = learning_rates[args.num_env]
-    value_coeff = value_coeffs[args.num_env]
-    batch_size = batch_sizes[args.num_env]
-    size_layer = size_layers[args.num_env]
+    if algo_name == 'SAC':
```
Comment (alxbilger) [inline rl.py:218]: Polymorphism must be used instead of multiple conditions

File: rl.py [hunk 8]
```diff
@@ -182,71 +320,133 @@ def sec_to_hours(seconds):
...
-               final_rewards.append(final_r)
-               rewards.append(r)
-               steps.append(timer_limit*20*(idx+1))
-
-
-               with open("./Results_benchmark/" +  name + "/rewards_"+id+".txt", 'w') as fp:
-                   json.dump([rewards, steps], fp)
-               with open("./Results_benchmark/" +  name + "/final_rewards_"+id+".txt", 'w') as fp:
-                   json.dump([final_rewards, steps], fp)
+    if TRAIN == 'new':
```
Comment (alxbilger) [inline rl.py:323]: Again polymorphism must be used. At minimum, it must be separated in 2 functions.

**Review by alxbilger — CHANGES_REQUESTED** [review]
Overall the script becomes too long, therefore difficult to read. Polymorphism could be a option to refactor parts of the script and improve the readability. Do not hesitate to separate the script into several.

File: rl.py (hunk 2 above)
Comment (samuelmyoussef) [reply rl.py:30]: It's a convention in gym to add the env version after the name, in this case version 0. If a significant change is made to the environment like changing how its reward is calculated for example, it would be appropriate to change it to v1.

File: README.md (hunk 3 above)
Comment (samuelmyoussef) [reply README.md]: linked a reference to the environments section where they are already listed
https://github.com/SofaDefrost/SofaGym/blob/a02a705c4d3b4fa88765b48c6ed5c0c602860349/README.md?plain=1#L114

File: README.md (hunk 4 above)
Comment (samuelmyoussef) [reply README.md]: included them
https://github.com/SofaDefrost/SofaGym/blob/a02a705c4d3b4fa88765b48c6ed5c0c602860349/README.md?plain=1#L115-L118

File: README.md (hunk 5 above)
Comment (samuelmyoussef) [reply README.md]: added all the parameters in the documentation

File: rl.py (hunk 6 above)
Comment (samuelmyoussef) [reply rl.py]: I moved this part to yaml files to save all the parameters of the model and the training, they can be changed by editing the yaml files directly

**Comment by samuelmyoussef** [comment]
I made modifications according to @alxbilger recommendations:
- Implemented classes for the RL agent to separate them according to the framework used:
  - SofaTestAgent: to create an agent only to test the env
  - SofaBaseAgent: the base class for the training classes
  - SB3Agent: to train the agent using Stable Baselines3
  - RlberryAgent: to train the agent using rlberry
- Added all model hyperparameters and training parameters into yaml files to be edited directly instead of parsing command line arguments.
- The `rl.py` script could now be used to create and train an RL agent by specifying the wanted environment, the algorithm, and the framework to use (SB3 or rlberry for now), and the interface is the same whichever the framework chosen.

**Review by alxbilger — CHANGES_REQUESTED** [review]
Could you just document the base classes (SofaBaseAgent and SofaTestAgent)?
